core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+            MessageEventTypes.NOTHING.name: self._do_nothing,
+        }
+
+    ##########################################################################
+    ###                            API SECTION                             ###
+    ##########################################################################
+
+    def _do_nothing(self, payload: dict):
+        """
+        Do nothing.
+        """
+        pass
+
+    def get_global_state(self):
+        """
+        Gets the global state which should be a list of dictionaries
+        """
+        timeout = time.time() + 5
+        tx_receipt = None
+        while time.time() < timeout:
+            try:
+                tx_receipt = requests.get(
+                    "http://localhost:{0}/state".format(self.port))
+                tx_receipt.raise_for_status()
+                if tx_receipt:
+                    break
+            except (UnboundLocalError, requests.exceptions.ConnectionError) as e:
+                logging.info("HTTP GET error, got: {0}".format(e))
+                continue
+        logging.info("global state:{}".format(tx_receipt.json()))
+        return tx_receipt.json()
+
+    def get_diffs(self, old_state: list, new_state: dict) -> list:
+        """
+        Iterate through oldState and newState to see any differences
+        Take action based on the differences
+        """
+        tx_diffs = [tx for tx in new_state.get('messages') if (
+                        tx not in old_state)]
+        return tx_diffs
+
+    def get_state_diffs(self, event_filter: object) -> list:
+        """
+        Gets state, then finds diffs, then sets state of blockchain.
+        """
+        new_state = self.get_global_state()
+        if new_state:
+            state_diffs = self.get_diffs(self.state, new_state)
+            filtered_diffs = [tx for tx in state_diffs if event_filter(tx)]
+            logging.info("filtered diffs:{}".format(filtered_diffs))
+            self.update_state(new_state)
+            return filtered_diffs
+
+    def check_malformed_content(self) -> None:
+        """
+        Checks whether content is of the form
+            `{
+                type: ... ,
+                payload: ...
+            }`
+        """
+        pass
+
+    def update_state(self, new_state_wrapper: list) -> None:
+        """
+        Given the freshly-downloaded state, call a handler on each transaction
+        that was not already present in our own state
+        """
+        new_state = dict(new_state_wrapper).get('messages')
+        len_state = len(self.state)
+        for i in new_state[len_state:]:
+            self.state.append(i)
+
+    def setter(self, key: str, value: object, flag: bool = False) -> str:
+        """
+        Provided a key and a JSON/np.array object, upload the object to
+        IPFS and then store the hash as the value on the blockchain. The key
+        should be a backward reference to a prior tx
```

Comment:
What's a tx and what do you mean with "backward reference"? Can we be more clear in the docs?

Reply:
both of these are elaborated on in Quip; function docstrings are not the place to explain blockchain state design semantics

Reply:
I'm resolving for now. for future reference tx is the common shorthand for transaction